src/database.py

```
@@ -1441,21 +1445,71 @@ async def _run_migrations(conn: aiosqlite.Connection) -> None:
+
+def _restore_database_file(backup_path: Path, db_path: Path) -> int:
+    """Validate *backup_path*, then atomically install it without stale WAL state."""
+    try:
+        with closing(sqlite3.connect(f"file:{backup_path}?mode=ro", uri=True)) as backup:
+            integrity = backup.execute("PRAGMA integrity_check").fetchone()
+            if integrity != ("ok",):
+                raise RuntimeError(f"backup integrity_check failed: {integrity!r}")
+            version = int(backup.execute("PRAGMA user_version").fetchone()[0])
+        restore_tmp = db_path.with_name(f".{db_path.name}.restore")
+        with closing(sqlite3.connect(f"file:{backup_path}?mode=ro", uri=True)) as source:
+            with closing(sqlite3.connect(restore_tmp)) as destination:
```

Ruff (0.16.3)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)